apps/src/lib/levelbuilder/lesson-editor/activitiesEditorRedux.js

```diff
@@ -0,0 +1,555 @@
+  activityPosition,
+  activitySectionPosition,
+  levelPosition,
+  levelId
+});
+
+export const setActiveVariant = (
+  activityPosition,
+  activitySectionPosition,
+  levelPosition,
+  id
+) => ({
+  type: SET_ACTIVE_VARIANT,
+  activityPosition,
+  activitySectionPosition,
+  levelPosition,
+  id
+});
+
+export const setField = (
+  activityPosition,
+  activitySectionPosition,
+  levelPosition,
+  modifier
+) => ({
+  type: SET_FIELD,
+  activityPosition,
+  activitySectionPosition,
+  levelPosition,
+  modifier
+});
+
+export const reorderLevel = (
+  activityPosition,
+  activitySectionPosition,
+  originalLevelPosition,
+  newLevelPosition
+) => ({
+  type: REORDER_LEVEL,
+  activityPosition,
+  activitySectionPosition,
+  originalLevelPosition,
+  newLevelPosition
+});
+
+export const moveLevelToActivitySection = (
+  activityPosition,
+  activitySectionPosition,
+  levelPosition,
+  newActivitySectionPosition
+) => ({
+  type: MOVE_LEVEL_TO_ACTIVITY_SECTION,
+  activityPosition,
+  activitySectionPosition,
+  levelPosition,
+  newActivitySectionPosition
+});
+
+export const addLevel = (activityPosition, activitySectionPosition, level) => ({
+  type: ADD_LEVEL,
+  activityPosition,
+  activitySectionPosition,
+  level
+});
+
+export const moveActivity = (activityPosition, direction) => ({
+  type: MOVE_ACTIVITY,
+  activityPosition,
+  direction
+});
+
+export const moveActivitySection = (
+  activityPosition,
+  activitySectionPosition,
+  direction
+) => ({
+  type: MOVE_ACTIVITY_SECTION,
+  activityPosition,
+  activitySectionPosition,
+  direction
+});
+
+export const removeActivity = activityPosition => ({
+  type: REMOVE_ACTIVITY,
+  activityPosition
+});
+
+export const removeActivitySection = (
+  activityPosition,
+  activitySectionPosition
+) => ({
+  type: REMOVE_ACTIVITY_SECTION,
+  activityPosition,
+  activitySectionPosition
+});
+
+export const setActivity = (
+  activitySectionPosition,
+  oldActivityPosition,
+  newActivityPosition
+) => ({
+  type: SET_ACTIVITY,
+  activitySectionPosition,
+  oldActivityPosition,
+  newActivityPosition
+});
+
+export const addTip = (activityPosition, activitySectionPosition, tip) => ({
+  type: ADD_TIP,
+  activityPosition,
+  activitySectionPosition,
+  tip
+});
+
+function updateActivityPositions(activities) {
+  for (let i = 0; i < activities.length; i++) {
+    activities[i].position = i + 1;
+  }
+}
+
+function updateActivitySectionPositions(activities) {
+  activities.forEach(activity => {
+    let position = 1;
+    activity.activitySections.forEach(activitySection => {
+      activitySection.position = position;
+      position++;
+    });
+  });
+}
+
+function updateLevelPositions(levels) {
+  for (let i = 0; i < levels.length; i++) {
+    levels[i].position = i + 1;
+  }
+}
+
+export const NEW_LEVEL_ID = -1;
+
+function activities(state = [], action) {
+  let newState = _.cloneDeep(state);
+
+  switch (action.type) {
+    case INIT:
+      return action.activities;
```

Comment:
In comments on https://github.com/code-dot-org/code-dot-org/pull/36627, I'm requesting the extraction of prop type "shapes". A lot of our redux code relies on the components which use it to do the type checking, however one idea would be to explicitly check against those proptypes here by calling https://github.com/facebook/prop-types#proptypescheckproptypes .